### PR TITLE
csprng multi backend fix

### PR DESCRIPTION
**Pull request by frkv**
Fixes issue where glued AES causes faulty behavior

-Adds fallback to use entropy output in entropy_cc310 driver
-Adds manifest update to take in nrfxlib pr #405 which sets the
 CTR_DRBG_CSPRNG_GENERATOR in case glue is enabled

manifest pr: https://github.com/nrfconnect/sdk-nrfxlib/pull/405

ref: NCSDK-8228

Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>

### CODE REVIEW

**Comment by eugmes** [comment]
I'm attempting to test this here: https://jenkins-ncs.nordicsemi.no/blue/organizations/jenkins/latest%2Fsub%2Ftest-fw-nrfconnect-ble/detail/master/7193/pipeline/206

**Comment by eugmes** [comment]
BLE test are still failing.

**Comment by eugmes** [comment]
The BLE tests are still failing. If this is not fixed by this PR in next few hours, I propose to merge #3898 first to avoid further breakage.

File: Kconfig.nrf
Comment (rugeGerritsen) [inline Kconfig.nrf:18]: No need to have `default n if BT`. The default value is `n`

Comment (frkv) [reply Kconfig.nrf:18]: No. The construct here (due to the order of parameters) will do default n for BT only if Zigbee or OpenThread is not set. This is intentional

File: Kconfig.nrf
Comment (rugeGerritsen) [inline Kconfig.nrf:21]: Why redefine the config ZIGBEE here? Shouldn't this select statement be put in nrf/subsys/zigbee?

File: Kconfig.nrf
Comment (rugeGerritsen) [inline Kconfig.nrf:23]: Similar comment as for ZIGBEE. Better to not create a new config

File: Kconfig.nrf
Comment (rugeGerritsen) [inline Kconfig.nrf:30]: What is the reason for adding `!(SPM || BUILD_WITH_TFM)` to this line? It is not explained in either the comment or the commit message. Also, these configs already select `ENTROPY_CC3XX ` by default.

File: Kconfig.nrf
Comment (rugeGerritsen) [inline Kconfig.nrf:27]: We could consider making this statement more future proof
```suggestion
# nRF Connect SDK needs a larger default stacks for certain configurations.
```

**Review by rugeGerritsen — CHANGES_REQUESTED** [review]
Initial proposal looks ok, but needs some cleanup

File: Kconfig.nrf
Comment (e-rk) [inline Kconfig.nrf:17]: I think a better option would be to use `NET_L2_OPENTHREAD` instead of `OPENTHREAD_NRF_SECURITY`. The latter can be set to either `y` or `n` and still have functional OpenThread stack, while the fromer definitely enables/disables the stack.

File: Kconfig.nrf
Comment (rugeGerritsen) [reply Kconfig.nrf:18]: I see. Why not add `if ZIGBEE || OPENTHREAD_NRF_SECURITY` at the place where `config ENTROPY_CC3XX` is declared? That removes duplication

File: Kconfig.nrf
Comment (e-rk) [inline Kconfig.nrf]: Isn't the `default` statement in line `17` enough? Is the `select` needed? I think Kconfig gives more meaningful errors in case of misconfiguration when `depends` is used instead.

**Comment by frkv** [comment]
Minor force-push. No need for select of ENTROPY_CC3XX for Zigbee/OpenThread. They can rely on the default y from the driver.

**Review by eugmes — COMMENTED** [review]
Most ble tests pass now. There is one newly failing mesh test though, it fails with the same symptoms. Not sure if that is related: https://jenkins-ncs.nordicsemi.no/blue/organizations/jenkins/latest%2Fsub%2Ftest-fw-nrfconnect-ble/detail/master/7198/pipeline

File: Kconfig.nrf
Comment (frkv) [reply Kconfig.nrf:17]: But does the NET_L2_OPENTHREAD imply usage of ENTROPY_CC3XX. From my understanding the need for OPENTHREAD_NRF_SECURITY is to override OPENTHREAD_MBEDTLS (built in mbed TLS)?

File: Kconfig.nrf
Comment (frkv) [reply Kconfig.nrf]: This has been removed

File: Kconfig.nrf
Comment (frkv) [reply Kconfig.nrf:18]: This has been discussed with one of the code-owners. The addition of subsystem knowledge on the level of the driver was rejected as a sub-optimal solution. Also the statement would look very hairy

something like this:
default y if !(BT & !(ZIGBEE || OPENTHREAD_NRF_SECURITY || SPM || BUILD_WITH_TFM))

It was decided that this would be better suited in the Kconfig.nrf scope

File: Kconfig.nrf
Comment (e-rk) [reply Kconfig.nrf:17]: Maybe my understanding is wrong, but doesn't setting the `ENTROPY_CC3XX` add the CC3XX generator as the default entropy source for all mbedTLS random number contexts, including both vanilla and nrf_security?

File: Kconfig.nrf
Comment (frkv) [reply Kconfig.nrf]: This has been removed. It was added in error

File: Kconfig.nrf
Comment (frkv) [reply Kconfig.nrf:24]: The SPM and BUILD_WITH_TFM will call into the Secure Partition Environment (secure domain in TrustZone). This means that the ENTROPY_CC3XX does not require having stack

Also note that in case of SPM / BUILD_WITH_TFM, the ENTROPY_CC3XX is enabled by default as the nRF9160 and nRF5340 does not contain a RNG device in the non-secure and would need to enable ENTROPY_CC3XX always

File: Kconfig.nrf
Comment (lmaciejonczyk) [reply Kconfig.nrf:17]: In theory someone could configure OpenThread to use built in mbed TLS but entropy from CC3XX.

File: Kconfig.nrf
Comment (frkv) [reply Kconfig.nrf:21]: Will be considered as a follow-up task once we get this feature in

Right now it is very explicit on what is going on for the sole purpose of being visible in case we rework this at a later time. There are other locations where the default stack size is manipulated in similar type Kconfig defaults (or defconfigs). Stating "certain configurations" makes it look like this is the place that all this activity will happen (which is not true)...

File: Kconfig.nrf
Comment (frkv) [reply Kconfig.nrf:17]: Unfortunately mbed TLS from pure mbed TLS would be using a different context-size than the one used by the CryptoCell. In this case SW-implemented (by original mbed  TLS) would need to be used

This will also be more inefficient as the CTR_DRBG will be implemented by software rather than being HW accelerated

Do note that there are some limitations in the mechanism of selecting which driver that supports CTR_DRBG in the CSPRNG (it is a simple choice right now). We are working on a proposal to fix this to ensure that ENTROPY_CC3XX can be used in this case, but it would require adding a whole new "ENTROPY_HAS_CSPRNG_FEATURE" to enable (or not) in driver configuration. This would need upstream approval in Zephyr

**Comment by frkv** [comment]
Looking at some DTS-related issues with CryptoCell being auto-selected on nRF52 devices... Will update when new run is possible

File: Kconfig.nrf
Comment (joerchan) [reply Kconfig.nrf:18]: I don't think it has to be hairy, to me it seems like this would do the exact same thing
```
drivers/entropy/Kconfig
@@ -10,7 +10,7 @@ config ENTROPY_CC3XX
-       default y
+       default y if !BT
subsys/zigbee/Kconfig
@@ -16,6 +16,7 @@ menuconfig ZIGBEE
+       imply ENTROPY_CC3XX
```
And adding imply ENTROPY_CC3XX the other places you mentioned, assuming it is imply and not select.

That doesn't look worse than splitting Kconfig definition into two, atleast using imply instead seems to handle "avoid subsystem knowledge in other places" argument.

**Comment by eugmes** [comment]
Note that ble_samples stage has being disabled in ble plan. This should make it green for now.

File: Kconfig.nrf
Comment (rugeGerritsen) [reply Kconfig.nrf:17]: I suggest to just remove this redeclaration of `ENTROPY_CC3XX`. In `nrf/drivers/entropy/Kconfig` we already set this config to `default y`. Adding `default y if ZIGBEE || OPENTHREAD_NRF_SECURITY` has no effect

File: Kconfig.nrf
Comment (joerchan) [reply Kconfig.nrf:24]: This would apply to all stacks that call into the entropy driver, right? I think moving all CC3XX handling that requires this this stack increase to a thread so that stack increase would only have to be once.
We avoided using the chosen entropy device in the `bt_rand` function call because it would affect all BT threads because of this.

File: Kconfig.nrf
Comment (frkv) [reply Kconfig.nrf:18]: I will prefer that @anangl or code-owners on generic driver support in nRF Connect SDK come with a decision on the placement on !BT in the driver

Imply on Zigbee was discussed in a call and other options. There are surely multiple ways to solve this, but the additions in this PR was cleanly stated as a **temporary fix** to try to get CI green.

We will address this issue later on when we have more time,  but probably in a different PR than this

I will remember to link to this, soon to be closed PR in any follow-up discussions

File: Kconfig.nrf
Comment (frkv) [reply Kconfig.nrf:24]: We have started speaking about putting the entire entropy-driver in its own thread with a proficiently sized thread to ensure that this does not impact users. 

This, however, is not a quick fix and will not happen in the coming release of nRF Connect SDK

**Comment by frkv** [comment]
Closed to be respawned in a different scope after the coming release

**Pull request closed without merging**
